File: packages/paperclip/native/src/pc/ast.rs
use crate::annotation::ast as annotation_ast;
use crate::base::ast::{BasicRaws, Range};
use crate::css::ast as css_ast;
use crate::js::ast as js_ast;
use serde::Serialize;
use std::fmt;
use std::str;

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct ElementRaws {
  pub before: String,
}

impl ElementRaws {
  pub fn new(before: &[u8]) -> ElementRaws {
    ElementRaws {
      before: str::from_utf8(before).unwrap().to_string(),
    }
  }
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct Element {
  pub id: String,
  pub raws: ElementRaws,

  pub range: Range,

  #[serde(rename = "openTagRange")]
  pub open_tag_range: Range,

  #[serde(rename = "tagNameRange")]
  pub tag_name_range: Range,

  #[serde(rename = "tagName")]
  pub tag_name: String,
  pub attributes: Vec<Attribute>,
  pub children: Vec<Node>,
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct Comment {
  pub id: String,
  pub raws: BasicRaws,
  pub range: Range,
  pub annotation: annotation_ast::Annotation,
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct ValueObject {
  pub id: String,
  pub range: Range,
  pub value: String,
  pub raws: BasicRaws,
}

#[derive(Debug, PartialEq, Serialize, Clone)]
#[serde(tag = "nodeKind")]
pub enum Node {
  Text(ValueObject),
  Comment(Comment),
  Element(Element),
  Fragment(Fragment),
  StyleElement(StyleElement),
  Slot(Slot),
}

#[derive(Debug, PartialEq, Serialize, Clone)]
#[serde(tag = "pcObjectKind")]
pub enum PCObject<'a> {
  Node(&'a Node),
  CSSObject(css_ast::CSSObject<'a>),
  JSObject(&'a js_ast::Expression),
}

impl<'a> PCObject<'a> {
  pub fn get_range(&'a self) -> &'a Range {
    match self {
      PCObject::Node(node) => node.get_range(),
      PCObject::CSSObject(css) => css.get_range(),
      PCObject::JSObject(js) => js.get_range(),
    }
  }
}

impl Node {
  pub fn get_range(&self) -> &Range {
    match self {
      Node::Text(value) => &value.range,
      Node::Comment(value) => &value.range,
      Node::Element(value) => &value.range,
      Node::Fragment(value) => &value.range,
      Node::StyleElement(value) => &value.range,
      Node::Slot(value) => &value.range,
    }
  }
  pub fn walk<F: FnMut(&Node) -> bool>(&self, each: &mut F) -> bool {
    if !(each)(self) {
      return false;
    }
    if let Some(children) = get_children(self) {
      for child in children {
        if !child.walk(each) {
          return false;
        }
      }
    }
    return true;
  }
  pub fn get_id(&self) -> &String {
    match self {
      Node::Text(value) => &value.id,
      Node::Comment(value) => &value.id,
      Node::Element(value) => &value.id,
      Node::Fragment(value) => &value.id,
      Node::StyleElement(value) => &value.id,
      Node::Slot(value) => &value.id,
    }
  }
  pub fn get_object_by_id<'a>(&self, id: &String) -> Option<PCObject> {
    if self.get_id() == id {
      return Some(PCObject::Node(self));
    }

    if let Node::Slot(slot) = self {
      return slot
        .script
        .get_object_by_id(id)
        .and_then(|object| match object {
          js_ast::JSObject::PCObject(obj) => Some(obj),
          js_ast::JSObject::Expression(expr) => match expr {
            js_ast::Expression::Node(node) => Some(PCObject::Node(node)),
            _ => Some(PCObject::JSObject(expr)),
          },
        });
    } else if let Node::StyleElement(style_element) = self {
      return style_element
        .sheet
        .get_object_by_id(id)
        .and_then(|object| Some(PCObject::CSSObject(object)));
    }

    get_children(self).and_then(|children| {
      for child in children {
        let nested = child.get_object_by_id(id);
        if nested != None {
          return nested;
        }
      }
      None
    })
  }
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct Slot {
  pub id: String,
  // !{slot}
  #[serde(rename = "omitFromCompilation")]
  pub omit_from_compilation: bool,
  pub script: js_ast::Expression,
  pub range: Range,
  pub raws: BasicRaws,
}

impl fmt::Display for Node {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    match self {
      Node::Text(text) => write!(f, "{}", &text.value),
      Node::Slot(slot) => write!(f, "{{{}}}", &slot.script.to_string()),
      Node::Comment(comment) => write!(f, "<!--[Annotation]-->"),
      Node::Fragment(node) => write!(f, "{}", node.to_string()),
      Node::Element(element) => write!(f, "{}", element.to_string()),
      Node::StyleElement(element) => write!(f, "{}", element.to_string()),
    }
  }
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct AttributeStringValue {
  pub id: String,
  pub value: String,
  pub range: Range,
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct AttributeSlotValue {
  pub id: String,
  pub script: js_ast::Expression,
  pub range: Range,
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct AttributeDynamicStringValue {
  pub id: String,
  pub values: Vec<AttributeDynamicStringPart>,
  pub range: Range,
}

impl fmt::Display for AttributeDynamicStringValue {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    for value in &self.values {
      match value {
        AttributeDynamicStringPart::Slot(stmt) => {
          stmt.fmt(f);
        }
        AttributeDynamicStringPart::ClassNamePierce(stmt) => {
          write!(f, "${}", stmt.class_name);
        }
        AttributeDynamicStringPart::Literal(stmt) => {
          write!(f, "{}", stmt.value);
        }
      }
    }
    Ok(())
    // match &self {
    //   AttributeValue::String(value) => write!(f, "{}", value.to_string()),
    //   AttributeValue::Slot(script) => write!(f, "{{{}}}", script.to_string()),
    //   AttributeValue::DyanmicString(script) => write!(f, "{{{}}}", script.to_string()),
    // }
  }
}

#[derive(Debug, PartialEq, Serialize, Clone)]
#[serde(tag = "partKind")]
pub enum AttributeDynamicStringPart {
  Literal(AttributeDynamicStringLiteral),
  ClassNamePierce(AttributeDynamicStringClassNamePierce),
  Slot(js_ast::Expression),
}

impl AttributeDynamicStringPart {
 pub fn get_id(&self) -> &String {
    match self {
      AttributeDynamicStringPart::ClassNamePierce(expr) => &expr.id,
      AttributeDynamicStringPart::Literal(expr) => &expr.id,
      AttributeDynamicStringPart::Slot(expr) => &expr.get_id()
    }
  }
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct AttributeDynamicStringLiteral {
  pub id: String,
  pub value: String,
  pub range: Range,
}

#[derive(Debug, PartialEq, Serialize, Clone)]
#[serde(tag = "className")]
pub struct AttributeDynamicStringClassNamePierce {
  pub id: String,
  #[serde(rename = "className")]
  pub class_name: String,
  pub range: Range,
}

impl fmt::Display for AttributeStringValue {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    write!(f, "\"{}\"", self.value)
  }
}

pub fn fmt_attributes(attributes: &Vec<Attribute>, f: &mut fmt::Formatter) -> fmt::Result {
  for attribute in attributes {
    write!(f, " {}", attribute.to_string())?;
  }
  Ok(())
}

pub fn fmt_start_tag<'a>(
  tag_name: &'a str,
  attributes: &Vec<Attribute>,
  f: &mut fmt::Formatter,
) -> fmt::Result {
  write!(f, "<{}", tag_name)?;
  fmt_attributes(attributes, f)?;
  write!(f, ">")?;
  Ok(())
}

pub fn fmt_end_tag<'a>(tag_name: &'a str, f: &mut fmt::Formatter) -> fmt::Result {
  write!(f, "</{}>", tag_name)?;
  Ok(())
}

impl fmt::Display for Element {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    fmt_start_tag(&self.tag_name.as_str(), &self.attributes, f)?;
    for child in &self.children {
      write!(f, "{} ", child.to_string())?;
    }
    fmt_end_tag(&self.tag_name.as_str(), f)?;
    Ok(())
  }
}

#[derive(Debug, PartialEq, Serialize, Clone)]
#[serde(tag = "attrKind")]
pub enum Attribute {
  ShorthandAttribute(ShorthandAttribute),
  SpreadAttribute(SpreadAttribute),
  KeyValueAttribute(KeyValueAttribute),
  PropertyBoundAttribute(PropertyBoundAttribute),
}

impl fmt::Display for Attribute {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    match self {
      Attribute::ShorthandAttribute(attr) => attr.fmt(f),
      Attribute::KeyValueAttribute(attr) => attr.fmt(f),
      Attribute::SpreadAttribute(attr) => attr.fmt(f),
      Attribute::PropertyBoundAttribute(attr) => attr.fmt(f),
    }
  }
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct SpreadAttribute {
  pub id: String,
  // !{...slot}
  #[serde(rename = "omitFromCompilation")]
  pub omit_from_compilation: bool,
  pub script: js_ast::Expression,
  pub range: Range,
}

impl fmt::Display for SpreadAttribute {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    write!(f, "{{...{}}}", self.script.to_string())
  }
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct ShorthandAttribute {
  pub id: String,
  pub reference: js_ast::Expression,
  pub range: Range,
}

impl ShorthandAttribute {
  pub fn get_name(&self) -> Result<&String, &'static str> {
    match &self.reference {
      js_ast::Expression::Reference(reference) => {
        if reference.path.len() == 1 {
          Ok(&reference.path[0].name)
        } else {
          Err("Unexpected Expression")
        }
      }
      _ => Err("Unexpected Expression"),
    }
  }
}

impl fmt::Display for ShorthandAttribute {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    write!(f, "{{{}}}", self.reference.to_string())
  }
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct PropertyBoundAttribute {
  pub id: String,
  #[serde(rename = "bindingName")]
  pub binding_name: String,
  pub name: String,
  pub range: Range,
  pub value: Option<AttributeValue>,
}

impl fmt::Display for PropertyBoundAttribute {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    write!(f, "{}:{}", self.name, self.binding_name)?;
    if self.value == None {
      Ok(())
    } else {
      write!(f, "={}", self.value.as_ref().unwrap().to_string())
    }
  }
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct KeyValueAttribute {
  pub id: String,
  pub name: String,
  pub range: Range,
  pub value: Option<AttributeValue>,
}

impl fmt::Display for KeyValueAttribute {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    write!(f, "{}", self.name)?;
    if self.value == None {
      Ok(())
    } else {
      write!(f, "={}", self.value.as_ref().unwrap().to_string())
    }
  }
}

#[derive(Debug, PartialEq, Serialize, Clone)]
#[serde(tag = "attrValueKind")]
pub enum AttributeValue {
  DyanmicString(AttributeDynamicStringValue),
  String(AttributeStringValue),
  Slot(AttributeSlotValue),
}

impl fmt::Display for AttributeValue {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    match &self {
      AttributeValue::String(value) => write!(f, "{}", value.to_string()),
      AttributeValue::Slot(value) => write!(f, "{{{}}}", value.script.to_string()),
      AttributeValue::DyanmicString(script) => write!(f, "{{{}}}", script.to_string()),
    }
  }
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct StyleElement {
  pub id: String,
  pub raws: ElementRaws,
  pub attributes: Vec<Attribute>,
  pub sheet: css_ast::Sheet,
  pub range: Range,
}

impl fmt::Display for StyleElement {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    fmt_start_tag("style", &self.attributes, f)?;
    write!(f, "{}", self.sheet.to_string())?;
    fmt_end_tag("style", f)?;
    Ok(())
  }
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct Fragment {
  pub id: String,
  pub range: Range,
  pub children: Vec<Node>,
}

impl fmt::Display for Fragment {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    write!(f, "")?;
    for child in &self.children {
      write!(f, "{}", child.to_string())?;
    }

    Ok(())
  }
}
pub fn get_children<'a>(expr: &'a Node) -> Option<&'a Vec<Node>> {
  match &expr {
    Node::Element(root) => Some(&root.children),
    Node::Fragment(root) => Some(&root.children),
    _ => None,
  }
}

pub fn get_imports<'a>(root_expr: &'a Node) -> Vec<&'a Element> {
  let mut imports = vec![];

  let children = get_children(root_expr);

  if children != None {
    for child in children.unwrap() {
      if let Node::Element(element) = &child {
        if element.tag_name == "import" && get_attribute_value("src", element) != None {
          imports.push(element);
        }
      }
    }
  }

  imports
}

pub fn get_tag_name<'a>(element: &'a Element) -> String {
  if element.tag_name.contains(".") {
    element.tag_name[..element.tag_name.find(".").unwrap()].to_string()
  } else {
    element.tag_name.to_string()
  }
}

pub fn get_tag_namespace<'a>(element: &'a Element) -> Option<String> {
  if element.tag_name.contains(".") {
    Some(element.tag_name[(element.tag_name.find(".").unwrap() + 1)..].to_string())
  } else {
    None
  }
}

pub fn get_parts<'a>(root_expr: &'a Node) -> Vec<&'a Element> {
  let mut parts = vec![];

  let children = get_children(root_expr);

  if children != None {
    for child in children.unwrap() {
      if let Node::Element(element) = &child {
        if has_attribute("component", element) {
          parts.push(element);
        }
      }
    }
  }

  parts
}

pub fn get_part_by_id<'a>(id: &String, root_expr: &'a Node) -> Option<&'a Element> {
  get_parts(root_expr)
    .iter()
    .find(|element| get_attribute_value("as", element) == Some(id))
    .map(|element| *element)
}

pub fn get_import_by_id<'a>(id: &String, root_expr: &'a Node) -> Option<&'a Element> {
  for import in get_imports(root_expr).iter() {
    if get_attribute_value("as", import) == Some(id) {
      return Some(import);
    }
  }
  None
}

pub fn get_import_by_src<'a>(src: &String, root_expr: &'a Node) -> Option<&'a Element> {
  for import in get_imports(root_expr).iter() {
    if get_attribute_value("src", import) == Some(src) {
      return Some(import);
    }
  }
  None
}

pub fn get_attribute<'a, 'b>(name: &'b str, element: &'a Element) -> Option<&'a Attribute> {
  for attribute in &element.attributes {
    if let Attribute::KeyValueAttribute(attr) = attribute {
      if attr.name == name {
        return Some(&attribute);
      }
    }
  }
  None
}

pub fn get_attribute_value<'a, 'b>(name: &'b str, element: &'a Element) -> Option<&'a String> {
  let attr = get_attribute(name, element);
  if let Some(att) = attr {
    if let Attribute::KeyValueAttribute(kv_attribute) = att {
      if let Some(expr) = &kv_attribute.value {
        if let AttributeValue::String(st) = &expr {
          return Some(&st.value);
        }
      }
    }
  }
  None
}

pub fn has_attribute<'a, 'b>(name: &'b str, element: &'a Element) -> bool {
  match get_attribute(name, element) {
    Some(attr) => true,
    None => false,
  }
}

pub fn get_import_ids<'a>(root_expr: &'a Node) -> Vec<&'a String> {
  let mut ids = vec![];
  for import in get_imports(root_expr) {
    if let Some(id) = get_attribute_value("as", &import) {
      ids.push(id);
    }
  }
  ids
}

pub fn get_import_identifier<'a>(import: &'a Element) -> Option<&'a String> {
  if let Some(id) = get_attribute_value("as", &import) {
    return Some(id);
  }
  if let Some(src) = get_attribute_value("src", &import) {
    return Some(src);
  }
  return None;
}

pub fn get_import_by_identifier<'a>(id: &String, root_expr: &'a Node) -> Option<&'a Element> {
  for import in get_imports(root_expr) {
    if get_attribute_value("as", &import) == Some(id)
      || get_attribute_value("src", &import) == Some(id)
    {
      return Some(import);
    }
  }
  return None;
}

pub fn get_part_ids<'a>(root_expr: &'a Node) -> Vec<&'a String> {
  let mut ids = vec![];

  // DEPRECATED
  for part in get_parts(root_expr) {
    if let Some(id) = get_attribute_value("as", &part) {
      ids.push(id);
    }
  }
  ids
}

File: packages/paperclip-utils/src/ast.ts
import { JsExpression, JsExpressionKind, Reference } from "./js-ast";
import {
  Sheet,
  traverseSheet,
  MixinRule,
  RuleKind,
  isRule,
  StyleExpression
} from "./css-ast";
import { BasicRaws, SourceLocation } from "./base-ast";
import * as crc32 from "crc32";
import { resolveImportFile } from "./resolve";
import * as path from "path";
import {
  LOGIC_TAG_NAME,
  DEFAULT_PART_ID,
  AS_ATTR_NAME,
  PREVIEW_ATTR_NAME
} from "./constants";
import { memoize } from "./memo";

export enum NodeKind {
  Fragment = "Fragment",
  Text = "Text",
  Annotation = "Annotation",
  Comment = "Comment",
  Element = "Element",
  StyleElement = "StyleElement",
  Slot = "Slot"
}

export type BaseNode<TKind extends NodeKind> = {
  nodeKind: TKind;
};

// TODO - include location here.
export type Text = {
  value: string;
  location: SourceLocation;
} & BaseNode<NodeKind.Text>;

export type Annotation = {
  properties: AnnotationProperty[];
  location: SourceLocation;
} & BaseNode<NodeKind.Annotation>;

export declare type Comment = {
  raws: BasicRaws;
  value: string;
  annotation: Annotation;
  location: SourceLocation;
} & BaseNode<NodeKind.Comment>;

export enum AnnotationPropertyKind {
  Text = "Text",
  Declaration = "Declaration"
}

type BaseAnnotationProperty<TKind extends AnnotationPropertyKind> = {
  kind: TKind;
};

export type TextAnnotation = {
  value: string;
  raws: BasicRaws;
} & BaseAnnotationProperty<AnnotationPropertyKind.Text>;

export type DeclarationAnnotation = {
  name: string;
  value: JsExpression;
  raws: BasicRaws;
} & BaseAnnotationProperty<AnnotationPropertyKind.Declaration>;

export type AnnotationProperty = TextAnnotation | DeclarationAnnotation;

export type ElementRaws = {
  before: string;
};

export type Element = {
  id: string;
  location: SourceLocation;
  raws: ElementRaws;

  // TODO - change this to OpenTag. Don't keep location here
  openTagLocation: SourceLocation;

  // TODO - change this to ElementTagName. name should go in value
  tagNameLocation: SourceLocation;
  tagName: string;
  attributes: Attribute[];
  value: string;
  children: Node[];
} & BaseNode<NodeKind.Element>;

export type StyleElement = {
  sheet: Sheet;
  location: SourceLocation;
  raws: BasicRaws;
} & BaseNode<NodeKind.StyleElement>;

export enum AttributeKind {
  ShorthandAttribute = "ShorthandAttribute",
  KeyValueAttribute = "KeyValueAttribute",
  SpreadAttribute = "SpreadAttribute",
  PropertyBoundAttribute = "PropertyBoundAttribute"
}

type BaseAttribute<TKind extends AttributeKind> = {
  attrKind: TKind;
};

type ShorthandAttribute = {
  reference: JsExpression;
  location: SourceLocation;
} & BaseAttribute<AttributeKind.ShorthandAttribute>;

type SpreadAttribute = {
  script: JsExpression;
  location: SourceLocation;
} & BaseAttribute<AttributeKind.SpreadAttribute>;

type KeyValueAttribute = {
  name: string;
  value?: AttributeValue;
  location: SourceLocation;
} & BaseAttribute<AttributeKind.KeyValueAttribute>;

export type PropertyBoundAttribute = {
  name: string;
  bindingName: string;
  value: AttributeValue;
  location: SourceLocation;
} & BaseAttribute<AttributeKind.PropertyBoundAttribute>;

export type Attribute =
  | ShorthandAttribute
  | SpreadAttribute
  | KeyValueAttribute
  | PropertyBoundAttribute;

export enum AttributeValueKind {
  DyanmicString = "DyanmicString",
  String = "String",
  Slot = "Slot"
}

export type BaseAttributeValue<TKind extends AttributeValueKind> = {
  attrValueKind: TKind;
};

export type StringAttributeValue = {
  value: string;
  location: SourceLocation;
} & BaseAttributeValue<AttributeValueKind.String>;

export enum DynamicStringAttributeValuePartKind {
  Literal = "Literal",
  ClassNamePierce = "ClassNamePierce",
  Slot = "Slot"
}

type BaseDynamicStringAttributeValuePart<
  TPartKind extends DynamicStringAttributeValuePartKind
> = {
  partKind: TPartKind;
};

type DynamicStringLiteralPart = {
  location: SourceLocation;
  value: string;
} & BaseDynamicStringAttributeValuePart<
  DynamicStringAttributeValuePartKind.Literal
>;

type DynamicStringClassNamePiercePart = {
  location: SourceLocation;
  className: string;
} & BaseDynamicStringAttributeValuePart<
  DynamicStringAttributeValuePartKind.ClassNamePierce
>;

type DynamicStringSlotPart = JsExpression &
  BaseDynamicStringAttributeValuePart<DynamicStringAttributeValuePartKind.Slot>;

export type DynamicStringAttributeValuePart =
  | DynamicStringLiteralPart
  | DynamicStringClassNamePiercePart
  | DynamicStringSlotPart;

export type DynamicStringAttributeValue = {
  values: DynamicStringAttributeValuePart[];
  location: SourceLocation;
} & BaseAttributeValue<AttributeValueKind.DyanmicString>;

export type SlotAttributeValue = {
  script: JsExpression;
  location: SourceLocation;
} & BaseAttributeValue<AttributeValueKind.Slot>;

export type AttributeValue =
  | StringAttributeValue
  | SlotAttributeValue
  | DynamicStringAttributeValue;

export type Fragment = {
  value: string;
  children: Node[];
  location: SourceLocation;
} & BaseNode<NodeKind.Fragment>;

export type Slot = {
  script: JsExpression;
  location: SourceLocation;
  raws: BasicRaws;
} & BaseNode<NodeKind.Slot>;

export type Node =
  | Text
  | Element
  | StyleElement
  | Fragment
  | Slot
  | Annotation
  | Comment;
export type Expression =
  | Node
  | Attribute
  | AttributeValue
  | StyleExpression
  | JsExpression
  | DynamicStringAttributeValuePart;

const a: AttributeValue = null;
export const getImports = (ast: Node): Element[] =>
  getChildrenByTagName("import", ast).filter(child => {
    return hasAttribute("src", child);
  });

export const getRelativeFilePath = fs => (
  fromFilePath: string,
  importFilePath: string
) => {
  const logicPath = resolveImportFile(fs)(fromFilePath, importFilePath);
  let relativePath = path.relative(path.dirname(fromFilePath), logicPath);
  if (relativePath.charAt(0) !== ".") {
    relativePath = `./${relativePath}`;
  }
  return relativePath;
};

export const getImportIds = (ast: Node): string[] =>
  getImports(ast)
    .map(node => getAttributeStringValue(AS_ATTR_NAME, node))
    .filter(Boolean) as string[];

export const getImportById = (id: string, ast: Node): Element | null =>
  getImports(ast).find(imp => {
    return getAttributeStringValue(AS_ATTR_NAME, imp) === id;
  });

export const getImportBySrc = (src: string, ast: Node): Element | null =>
  getImports(ast).find(imp => {
    return getAttributeStringValue("src", imp) === src;
  });

export const getChildren = (ast: Node): Node[] => {
  if (ast.nodeKind === NodeKind.Element || ast.nodeKind === NodeKind.Fragment) {
    return ast.children;
  }
  return [];
};

export const getStyleScopeId = (filePath: string) => {
  if (filePath.indexOf("file://") !== 0) {
    filePath = "file://" + filePath;
  }
  return crc32(filePath);
};

export const getChildrenByTagName = (tagName: string, parent: Node) =>
  getChildren(parent).filter(child => {
    return child.nodeKind === NodeKind.Element && child.tagName === tagName;
  }) as Element[];

export const findByNamespace = (
  namespace: string,
  current: Node,
  allChildrenByNamespace: Element[] = []
) => {
  if (current.nodeKind === NodeKind.Element) {
    if (current.tagName.split(".")[0] === namespace) {
      allChildrenByNamespace.push(current);
    }
  }
  for (const child of getChildren(current)) {
    findByNamespace(namespace, child, allChildrenByNamespace);
  }

  if (current.nodeKind === NodeKind.Element) {
    for (const attribute of current.attributes) {
      if (
        attribute.attrKind === AttributeKind.KeyValueAttribute &&
        attribute.value
      ) {
        if (
          attribute.value.attrValueKind === AttributeValueKind.Slot &&
          attribute.value.script.jsKind === JsExpressionKind.Node
        ) {
          findByNamespace(
            namespace,
            attribute.value.script,
            allChildrenByNamespace
          );
        }
      }
    }
  }

  return allChildrenByNamespace;
};

export const getMetaValue = (name: string, root: Node) => {
  const metaElement = getChildrenByTagName("meta", root).find(
    meta =>
      hasAttribute("src", meta) &&
      getAttributeStringValue("name", meta) === name
  );
  return metaElement && getAttributeStringValue("content", metaElement);
};

export const getAttribute = (name: string, element: Element) =>
  element.attributes.find(attr => {
    return (
      attr.attrKind === AttributeKind.KeyValueAttribute && attr.name === name
    );
  }) as KeyValueAttribute;

export const getAttributeValue = (name: string, element: Element) => {
  const attr = getAttribute(name, element);
  return attr && attr.value;
};

export const getAttributeStringValue = (name: string, element: Element) => {
  const value = getAttributeValue(name, element);
  return (
    value && value.attrValueKind === AttributeValueKind.String && value.value
  );
};

export const getStyleElements = (ast: Node): StyleElement[] => {
  const styleElements: StyleElement[] = [];

  traverseExpression(ast, (node: Node) => {
    if (node.nodeKind === NodeKind.StyleElement) {
      styleElements.push(node);
    }
  });

  return styleElements;
};

export const isVisibleElement = (ast: Element): boolean => {
  return !/^(import|logic|meta|style|part|preview)$/.test(ast.tagName);
};
export const isVisibleNode = (node: Node): boolean =>
  node.nodeKind === NodeKind.Text ||
  node.nodeKind === NodeKind.Fragment ||
  node.nodeKind === NodeKind.Slot ||
  (node.nodeKind === NodeKind.Element && isVisibleElement(node));

export const getVisibleChildNodes = (ast: Node): Node[] =>
  getChildren(ast).filter(isVisibleNode);

export const isComponent = (node: Node): node is Element =>
  node.nodeKind === NodeKind.Element &&
  hasAttribute("component", node) &&
  hasAttribute(AS_ATTR_NAME, node);

export const getParts = (ast: Node): Element[] =>
  getChildren(ast).filter(isComponent) as Element[];

export const getPartIds = (ast: Node): string[] =>
  getParts(ast)
    .map(node => getAttributeStringValue(AS_ATTR_NAME, node))
    .filter(Boolean) as string[];

export const getDefaultPart = (ast: Node): Element =>
  getParts(ast).find(
    part => getAttributeStringValue(AS_ATTR_NAME, part) === DEFAULT_PART_ID
  );

export const getLogicElement = (ast: Node): Element | null => {
  return getChildren(ast).find(
    child =>
      child.nodeKind === NodeKind.Element && child.tagName === LOGIC_TAG_NAME
  ) as Element;
};

export const hasAttribute = (name: string, element: Element) =>
  getAttribute(name, element) != null;

// https://github.com/crcn/tandem/blob/10.0.0/packages/common/src/state/tree.ts#L137
export const flattenTreeNode = memoize((current: Node): Node[] => {
  const treeNodeMap = getTreeNodeMap(current);
  return Object.values(treeNodeMap) as Node[];
});

export const getNodePath = memoize((node: Node, root: Node) => {
  const map = getTreeNodeMap(root);
  for (const path in map) {
    const c = map[path];
    if (c === node) return path;
  }
});

// TODO
export const getParentNode = (node: Node, root: Node) => {
  const nodePath = getNodePath(node, root).split(".");
  nodePath.pop();
  const map = getTreeNodeMap(root);
  return map[nodePath.join(".")] as Fragment | Element;
};

export const getTreeNodeMap = memoize(
  (current: Node, path = "0"): Record<string, Node> => {
    const map = {
      [path]: current
    };
    if (
      current.nodeKind === NodeKind.Fragment ||
      current.nodeKind === NodeKind.Element
    ) {
      Object.assign(
        map,
        ...current.children.map((child, i) =>
          getTreeNodeMap(child, path + "." + i)
        )
      );
    }
    return map;
  }
);

export const isComponentInstance = (
  node: Node,
  importIds: string[]
): node is Element => {
  return (
    node.nodeKind === NodeKind.Element &&
    importIds.indexOf(node.tagName.split(".").shift()) !== -1
  );
};

const maybeAddReference = (
  stmt: JsExpression,
  _statements: [Reference, string][] = []
) => {
  if (stmt.jsKind === JsExpressionKind.Reference) {
    _statements.push([stmt, null]);
  }
};

export const getMixins = (ast: Node): Record<string, MixinRule> => {
  const styles = getStyleElements(ast);
  const mixins: Record<string, MixinRule> = {};
  for (const style of styles) {
    traverseSheet(style.sheet, rule => {
      if (rule && isRule(rule) && rule.ruleKind === RuleKind.Mixin) {
        mixins[rule.name.value] = rule;
      }
    });
  }

  return mixins;
};

export const isNode = (ast: Expression): ast is Node =>
  NodeKind[(ast as Node).nodeKind] != null;
export const isAttribute = (ast: Expression): ast is Attribute =>
  AttributeKind[(ast as Attribute).attrKind] != null;
export const isAttributeValue = (ast: Expression): ast is AttributeValue =>
  AttributeValueKind[(ast as AttributeValue).attrValueKind] != null;

export const isJsExpression = (ast: Expression): ast is JsExpression =>
  JsExpressionKind[(ast as JsExpression).jsKind] != null;
export const isDynamicStringAttributeValuePart = (
  ast: Expression
): ast is DynamicStringAttributeValuePart =>
  DynamicStringAttributeValuePartKind[
    (ast as DynamicStringAttributeValuePart).partKind
  ] != null;

export const traverseExpression = (
  ast: Expression,
  each: (node: Expression) => void | boolean
) => {
  if (each(ast) === false) {
    return false;
  }
  if (isNode(ast)) {
    switch (ast.nodeKind) {
      case NodeKind.Element: {
        return (
          traverseExpressions(ast.attributes, each) &&
          traverseExpressions(ast.children, each)
        );
      }
      case NodeKind.Fragment: {
        return traverseExpressions(ast.children, each);
      }
      case NodeKind.StyleElement: {
        return traverseSheet(ast.sheet, each);
      }
    }
  }
  return true;
};

export const getCompletionItems = (root: Expression, position: number) => {
  let parent: Expression;
  let previousSibling: Expression;
  traverseExpression(root, expr => {
    if (!expr.location) {
      console.error("ERRRR", expr);
    }
  });
};

const traverseExpressions = (
  expressions: Expression[],
  each: (node: Expression) => void | boolean
) => {
  for (const child of expressions) {
    if (!traverseExpression(child, each)) {
      return false;
    }
  }
  return true;
};
export const getNestedReferences = (
  node: Node,
  _statements: [Reference, string][] = []
): [Reference, string][] => {
  if (node.nodeKind === NodeKind.Slot) {
    maybeAddReference(node.script, _statements);
  } else {
    if (node.nodeKind === NodeKind.Element) {
      for (const attr of node.attributes) {
        if (
          attr.attrKind == AttributeKind.KeyValueAttribute &&
          attr.value &&
          attr.value.attrValueKind === AttributeValueKind.Slot
        ) {
          if (attr.value.script.jsKind === JsExpressionKind.Node) {
            getNestedReferences(attr.value.script, _statements);
          } else if (attr.value.script.jsKind === JsExpressionKind.Reference) {
            _statements.push([attr.value.script, attr.name]);
          }
        } else if (
          attr.attrKind === AttributeKind.ShorthandAttribute &&
          attr.reference.jsKind === JsExpressionKind.Reference
        ) {
          _statements.push([attr.reference, attr.reference[0]]);
        } else if (
          attr.attrKind === AttributeKind.SpreadAttribute &&
          attr.script.jsKind === JsExpressionKind.Reference
        ) {
          _statements.push([attr.script, attr.script[0]]);
        }
      }
    }

    for (const child of getChildren(node)) {
      if (
        child.nodeKind === NodeKind.Element &&
        hasAttribute(PREVIEW_ATTR_NAME, child)
      ) {
        continue;
      }
      getNestedReferences(child, _statements);
    }
  }
  return _statements;
};
